File: 动态规划/找零钱.cpp
#include <iostream>
using namespace std;
#include <climits>
#include <vector>

class Solution
{
    vector<int> arr;
    int base;

public:
    /**
     * 最少货币数
     * @param arr int整型vector the array
     * @param aim int整型 the target
     * @return int整型
     */
    int minMoney(vector<int> &arr, int aim)
    {
        // write code here
        
        this->arr = arr;
        this->base = INT_MAX;
        for (int i : this->arr)
        {
            if (this->base > i)
            { // 直接选了
                this->base = i;
            }
        }
        return this->minMoney(aim);
    }
    int minMoney(int aim)
    {
        if (aim == 0)
        {
            return 0
        }
        
        // 递归基
        if (aim < this->base)
        { // 我一定不选这个
            return -1;
        }

        for (int i : this->arr)
        {
            if (aim == i)
            { // 直接选了
                return 1;
            }
        }
        // 找出最小的分支
        int min = INT_MAX;
        for (int i : arr)
        {
            int tmp = minMoney(aim - i); // 害得递归
            if (tmp < min)
            {
                min = tmp;
            }
        }
        // 返回分支+1
        return min + 1;
    }
};

int main(int argc, char const *argv[])
{
    Solution *s = new Solution();

    // vector<int> arr{25, 20, 5, 1};
    // int aim = 41;

    vector<int> arr{5,2,3};
    int aim = 20;
    int res = s->minMoney(arr, aim);
    cout << res << endl;
    return 0;
}

File: 动态规划/找零钱.py

#
# 代码中的类名、方法名、参数名已经指定，请勿修改，直接返回方法规定的值即可
#
# 最少货币数
# @param arr int整型一维数组 the array
# @param aim int整型 the target
# @return int整型
#
class Solution:
    def minMoney(self , arr, aim: int) -> int:
        # write code here
        # 找到最小面值
        if len(arr) == 0:
            return -1
        if (aim == 0):
            return 0
        self.base = min(arr)
        self.arr = arr
        self.memory = {}

        res = self.__minMoney(aim)
        if res == float("inf"):
            return -1
        return res

    def __minMoney(self, aim : int):
        if (aim == 0):
            return 0
        # 递归基, 说明不要了
        if(aim < self.base):
            return -1
        # 查询记忆
        if(self.memory.get(aim)):
            return self.memory[aim]
        # 向下搜索
        min = float("inf")
        for i in self.arr:
            tmp = self.__minMoney(aim - i)
            if (tmp < min and tmp != -1):
                min = tmp
        self.memory[aim] = min + 1
        return min + 1


if __name__ == "__main__":
    s = Solution()
    res = s.minMoney([2887,2355,335,4472,4561,1819,3808],4953)
    print(res)
    pass